2x speedup

RickytheGuy opened the pull request:
In the quest to generate global floodmaps, we've worked on making this go faster. Was able to get average timings from 2.1 sec to 0.8 sec for a 30 meter 1x1 degree DEM tile without bathmetry, 4.3 sec to 2.3 sec with bathymetry.

Changes:

- Numba signatures on some functions help load in much quicker
- pyarrow dependency to speed up CSV/parquet reading
- Improved weightbox creation speed
- Several edits for readability + memory consumption improve performance (such as using np.uint8 instead of np.float arrays)

I have tests to verify that both the burned DEM and the floodmaps generated with this pull request are identical.

RickytheGuy committed:
3.5x speedup

RickytheGuy committed:
force inputs of vdt interpolation to certain dtypes

RickytheGuy committed:
revert back to float64

RickytheGuy committed:
fixed error

RickytheGuy committed:
bathymetry optimizations

RickytheGuy committed:
signature on smoothing functions for smaller speedup

jlgutenson commented:
Thanks for sharing these updates Ricky!

Apologies for the delay!

Does this also contain the previous pull request updates you sent over?

RickytheGuy commented:
@jlgutenson This pull request is only edits to the most recent commit here (https://github.com/MikeFHS/curve2flood/commit/3c4ad077f4da5f179df2e03ff499aff0f15c5abb)

jlgutenson commented:
Sorry for the confusion @RickytheGuy! 

I mixed up the ARC updates you pushed with these for Curve2Flood.

I'm testing these results and had to make some minor tweaks to get your updates running on my test cases. I'm getting slightly different results using the same streamflow data. 

Here are the changes I made. 

1. Changed

`ARBathyMask = np.zeros((nrows+2,ncols+2), dtype=np.bool)
`
to 

`ARBathyMask = np.zeros((nrows+2,ncols+2), dtype=np.bool_)
`
2. Changed

```
COMID_Unique_TW = create_numba_dict_from(np.asarray(list(COMID_Unique_TW.keys())), np.asarray(list(COMID_Unique_TW.values())))
COMID_Unique_Depth = create_numba_dict_from(np.asarray(list(COMID_Unique_Depth.keys())), np.asarray(list(COMID_Unique_Depth.values())))
```

to 

```
keys_tw  = np.asarray(list(COMID_Unique_TW.keys()), dtype=np.int64)
vals_tw  = np.asarray(list(COMID_Unique_TW.values()), dtype=np.float64)
keys_dep = np.asarray(list(COMID_Unique_Depth.keys()), dtype=np.int64)
vals_dep = np.asarray(list(COMID_Unique_Depth.values()), dtype=np.float64)

COMID_Unique_TW    = create_numba_dict_from(keys_tw,  vals_tw)
COMID_Unique_Depth = create_numba_dict_from(keys_dep, vals_dep)
```

Do those changes impact your results or do your Curve2Flood outputs look the same?

RickytheGuy commented:
@jlgutenson Gotcha, implementing your changes still gives me the same output for my test case. The difference you saw must be coming from using source data with a different dtype.

RickytheGuy committed:
no need to delete first entry because only get unique values from where B > 0

RickytheGuy commented:
@jlgutenson I had implemented a way to get unique COMIDs quicker, but forgot to remove a line which deletes the first index (which is no longer necessary). This may result in a slight difference although the effect of 1 linkno removed did not change my floodmaps. I've fixed that in the above commit

jlgutenson commented:
Thanks for testing that @RickytheGuy!

I think you were not far off on the cause. 

I made changed the dtype from float64 to float32 in your `create_weightbox` function and that solved the issue for me. How does using the modified function below work for you? 

```
@njit("float32[:,:](int64, float32, float32)", cache=True)
def create_weightbox(tw: int, dx: float, dy: float):
    # tw is the number of cells in the top-width

    # --- Vectorized WeightBox creation ---
    n = 2*tw + 1
    # Create an array of indices [0, 1, ..., n-1]
    indices = np.arange(n)
    # Compute offsets from the center (center index is TW)
    # These offsets represent the "cell distance" in each direction.
    X = ((indices - tw) * dx) ** 2 # shape (n,)
    Y = ((indices - tw) * dy) ** 2 # shape (n,)
    
    # Broadcast to compute the squared distances:
    # For every cell, z2 = (dx * (x offset))^2 + (dy * (y offset))^2.
    # We use broadcasting: the row vector (X) and column vector (Y) combine to form an (n x n) array.
    WeightBox = X[None, :] + Y[:, None]  # shape (n, n)
    # Avoid very small values (to prevent division by zero)
    WeightBox = np.clip(WeightBox, 0.0001, None)
    WeightBox = 1.0 / WeightBox


    return (WeightBox).astype(np.float32)
```

RickytheGuy commented:
@jlgutenson Using float32 gives me an essentially identical floodmap, but with 0.0057% different cells than my base case. The reason why is the original code to the function looks like this:

```python
@njit(cache=True)
def CreateWeightAndElipseMask(TW_temp, dx, dy, TW_MultFact):
    TW = int(TW_temp)  #This is the number of cells in the top-width
     WeightBox = np.zeros((int(TW*2+1),int(TW*2+1)), dtype=np.float32)  #2D Array
```
Where weightbox is set to be float32. However, later on in the function you redefine weightbox:
```python
  n = 2*TW + 1
  # Create an array of indices [0, 1, ..., n-1]
  indices = np.arange(n)
  # Compute offsets from the center (center index is TW)
  # These offsets represent the "cell distance" in each direction.
  Y = indices - TW  # shape (n,)
  X = indices - TW  # shape (n,)
  # Broadcast to compute the squared distances:
  # For every cell, z2 = (dx * (x offset))^2 + (dy * (y offset))^2.
  # We use broadcasting: the row vector (X) and column vector (Y) combine to form an (n x n) array.
  z2 = (X * dx)**2  # shape (n,)
  z2 = z2[None, :] + ((Y * dy)**2)[:, None]  # shape (n, n)
  # Avoid very small values (to prevent division by zero)
  z2 = np.where(z2 < 0.0001, 0.0001, z2)
  WeightBox = 1.0 / z2
```
And here weightbox ends up being float64 (since z2 is float64 by default). The difference is all just due to differences in precision. I am fine with what you implemented, just letting you know that the original weightbox function has been returning a float64 array.

jlgutenson commented:
Thanks for pointing that out Ricky! I made a few additional changes on top of your and will push those up to the main branch shortly.

I think we can get by with using mainly float32 and int32. If you see something that make you think otherwise, just let me know!
